Source/MonoGame.Extended/Collections/ObjectPool.cs

```diff
@@ -0,0 +1,205 @@
+
+        public ObjectPool.Diagnostics Diagnostics { get; } = new ObjectPool.Diagnostics();
+
+        public int AvailableCount
+        {
+            get { return _objects.Count; }
+        }
+
+        public int MinimumSize
+        {
+            get { return _minimumSize; }
+            set
+            {
+                CheckLimits(value, _maximumSize);
+                _minimumSize = value;
+                AdjustSize();
+            }
+        }
+
+        public int MaximumSize
+        {
+            get { return _maximumSize; }
+            set
+            {
+                CheckLimits(_minimumSize, value);
+                _maximumSize = value;
+                AdjustSize();
+            }
+        }
+
+        public Func<T> FactoryMethod { get; }
+
+        public ObjectPool(int minimumSize = DefaultPoolMinimumSize, int maximumSize = DefaultPoolMaximumSize, Func<T> factoryMethod = null)
+        {
+            CheckLimits(minimumSize, maximumSize);
+            FactoryMethod = factoryMethod;
+            _maximumSize = maximumSize;
+            _minimumSize = minimumSize;
+            AdjustSize();
+        }
+
+        private void CheckLimits(int minimumSize, int maximumSize)
+        {
+            if (minimumSize < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(minimumSize), minimumSize, "Minimum pool size must be greater or equal to zero.");
+            }
+
+            if (maximumSize <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maximumSize), maximumSize, "Maximum pool size must be greater than zero.");
+            }
+
+            if (minimumSize > maximumSize)
+            {
+                throw new ArgumentOutOfRangeException(nameof(minimumSize), minimumSize, "Minimum pool size must be less than or equal to the maximum pool size.");
+            }
+        }
+
+        private void AdjustSize()
+        {
+            if (Interlocked.CompareExchange(ref _adjustSizeInProgress, 1, 0) != 0)
+            {
+                return;
+            }
+
+            while (_createdObjects.Count < MinimumSize)
+            {
+                _objects.Enqueue(CreateObject());
+            }
+
+            while (_createdObjects.Count > MaximumSize)
+            {
+                T poolable;
+                if (!_objects.TryDequeue(out poolable))
+                {
+                    continue;
+                }
+
+                Diagnostics.DecrementInstancesInUse();
+                Diagnostics.IncrementInstancesOverflow();
+                DestroyObject(poolable);
+            }
+
+            _adjustSizeInProgress = 0;
+        }
+
+        private T CreateObject()
+        {
+            T newObject;
+
+            if (FactoryMethod != null)
+            {
+                newObject = FactoryMethod();
+            }
+            else
+            {
+                newObject = (T)Activator.CreateInstance(typeof(T));
+            }
```

Comment:
Looks like it's currently limited to objects that have a parameterless constructor? One way we could get around this is by injecting a creation function `Func<T> createInstanceFunc` maybe.